OSDOCS-20042: adds OIDC authentication RHCL (#112868)
openshift/openshift-docs: modules/proc-rhcl-oidc-rbac-verify.adoc

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-oidc-rbac-verify_{context}"]
= Verifying RBAC with OIDC

[role="_abstract"]
You can verify that the role-based-access you configured is working your OpenID Connect (OIDC) authentication by conducting tests.

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.
* You have access to external or on-premise DNS.
* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You configured an `AuthPolicy` CR with role-based-access.

.Procedure

. Get an access token from within the cluster for the user John, who is assigned to the `member` role:
+
[source,terminal]
----

> $ ACCESS_TOKEN=$(oc run token --attach --rm --restart=Never -q --image=curlimages/curl -- http://keycloak.keycloak.svc.cluster.local:8080/realms/kuadrant/protocol/openid-connect/token -s -d 'grant_type=password' -d 'client_id=demo' -d 'username=john' -d 'password=p' -d 'scope=openid' | jq -r .access_token)

Reviewer: Same as https://github.com/openshift/openshift-docs/pull/112868/changes#r3383037255

* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You configured an `AuthPolicy` CR with role-based-access.

.Procedure

. Get an access token from within the cluster for the user John, who is assigned to the `member` role, by running the following command:
+
[source,terminal]
----
$ ACCESS_TOKEN=$(oc run token --attach --rm --restart=Never -q --image=curlimages/curl -- https://keycloak.keycloak.svc.cluster.local:8080/realms/kuadrant/protocol/openid-connect/token -s -d 'grant_type=password' -d 'client_id=demo' -d 'username=john' -d 'password=p' -d 'scope=openid' | jq -r .access_token)
----
+
[NOTE]
====
If your {keycloak} server is reachable from outside the cluster, you can get the token directly. Make sure that the `hostname` set in the OIDC issuer endpoint configured in the `AuthPolicy` CR matches the one used to get the token and is reachable from within the cluster.
====

. Retrieve your cluster base domain and assign it to an environment variable by running the following command:
+
[source,terminal]
----
$ export CLUSTER_DOMAIN=$(oc get ingress.config/cluster -o jsonpath='{.spec.domain}')
----

> . As John, send a `GET` request to **/resources** by running the following command:

Reviewer 2: Should a reference like this be in backticks instead, since this is an element that's in the code block itself?

```suggestion
. As John, send a `GET` request to `/resources` by running the following command:
```

If it is supposed to be, please also make the same changes throughout the procedure

+
[source,terminal,subs="+quotes"]
----
$ curl -H "Authorization: Bearer $ACCESS_TOKEN" https://api.$CLUSTER_DOMAIN/resources -i
----
+
.Example output
[source,text]
----
HTTP/1.1 200 OK
----

> . As John, send a `DELETE` request to **/resources/123** by running the following command:

Reviewer 2: What if my name isn't John?

JK, but while this probably isn't an issue for users reading this, maybe this could use the same formatting as the bullet list in the intro?

```suggestion
. As user `john`, send a `DELETE` request to **/resources/123** by running the following command:
```

Up to you. But if you make this change, please apply this to the rest of the procedure as well.

// Module included in the following assemblies:
//
// *deployment/rhcl-oids-authentication.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-oidc-rbac-verify_{context}"]
= Verifying RBAC with OIDC

[role="_abstract"]
You can verify that the role-based-access you configured is working your OpenID Connect (OIDC) authentication by conducting tests with different user roles.

The following example assumes the following setup:

* A {keycloak} instance running in the keycloak namespace listening at port 8080.
* A `kuadrant` {keycloak} realm.
* An OAuth2 client with Direct Access grant enabled.
* A user `john` who is assigned the `member` role and has the password `p`.

> * A user `jane` who is assigned the `member` and admin `roles` and has the  password `p`.

Reviewer: nit:

```suggestion
* A user `jane` who is assigned the `member` and `admin` roles and has the  password `p`.
```

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.

> * You have access to external or on-premise DNS.

Reviewer 3: No need for DNS

* You configured an `AuthPolicy` CR with role-based-access.

.Procedure

. Get an access token from within the cluster for the user John, who is assigned to the `member` role, by running the following command:
+
[source,terminal]
----
$ ACCESS_TOKEN=$(oc run token --attach --rm --restart=Never -q --image=curlimages/curl -- http://keycloak.keycloak.svc.cluster.local:8080/realms/kuadrant/protocol/openid-connect/token -s -d 'grant_type=password' -d 'client_id=demo' -d 'username=john' -d 'password=p' -d 'scope=openid' | jq -r .access_token)
----
+
[NOTE]
====
If your {keycloak} server is reachable from outside the cluster, you can get the token directly. Make sure that the `hostname` set in the OIDC issuer endpoint configured in the `AuthPolicy` CR matches the one used to get the token and is reachable from within the cluster.
====

. Retrieve your cluster base domain and assign it to an environment variable by running the following command:
+
[source,terminal]
----
$ export CLUSTER_DOMAIN=$(oc get ingress.config/cluster -o jsonpath='{.spec.domain}')
----

. As John, send a `GET` request to **/resources** by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ curl -H "Authorization: Bearer $ACCESS_TOKEN" http://api.$CLUSTER_DOMAIN/resources -i

Reviewer 3: Not sure here, should not we use https:// in our examples throughout official guides?

> $ curl -H "Authorization: Bearer $ACCESS_TOKEN" http://api.apps.$CLUSTER_DOMAIN/resources -i

Reviewer: For consistency with the other examples:

```suggestion
$ curl -H "Authorization: Bearer $ACCESS_TOKEN" http://api.$CLUSTER_DOMAIN/resources -i
```

----
+
.Example output
[source,text]
----
HTTP/1.1 200 OK
----

. As John, send a `DELETE` request to **/resources/123** by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ curl -H "Authorization: Bearer $ACCESS_TOKEN" -X DELETE http://api.$CLUSTER_DOMAIN/resources/123 -i
----
+
.Example output
[source,text]
----
HTTP/1.1 403 Forbidden
----

. As John, send a `GET` request to **/admin/settings** by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ curl -H "Authorization: Bearer $ACCESS_TOKEN" http://api.$CLUSTER_DOMAIN/admin/settings -i
----
+
.Example output
[source,text]
----
HTTP/1.1 403 Forbidden
----

> . Get an access token from within the cluster for the user Jane, who is assigned to the 'member' and 'admin' roles by running the following command:

Reviewer 4: ```suggestion
. Get an access token from within the cluster for the user Jane, who is assigned to the `member` and `admin` roles by running the following command:
```
